print()/cat() (#37)
r-devel/cran-cookbook: website/code_issues.qmd

-----------------------------------------------------------

# Using print()/cat()

## Problem

> You are using functions, like print() or cat(), to print unsuppressable output to the console.

Reviewer: I would use code syntax when referencing the functions:

```suggestion
You are using functions, like `print()` or `cat()`, to print unsuppressable output to the console.
```

Reviewer 2: 👍

## Solution

> Change print()/cat() to message(), warning(), stop(), or wrap them in if(verbose){}.

Reviewer: ```suggestion
Change `print()`/`cat()` to `message()`, `warning()`, `stop()`, or wrap them in `if(verbose){}`.
```

### Details

> Information messages, like loop counts or status updates, can clutter the console. While some users prefer this display, others appreciate less information on their console. The use of printing functions for console output is not forbidden on CRAN. However, this output must be suppressable by users.

Reviewer: I would comment on when to use `print()` or `cat()` functions. Like using `cat()` for print methods of classes. (I am not actually sure when I would use `print()` in a package.)

Reviewer 2: It would be useful to put a code example of how to implement the verbose usage or maybe link to an existing example like [here](https://www.educative.io/answers/what-is-verbose-in-r)
```
if(verbose){

}
```

Reviewer 2: scratch this comment - i see the example you added below 😄

:::{.callout-note}

Printing in special functions like print, summary, interactive functions or methods for generic functions is accepted. 

:::

To allow users to suppress the console output CRAN recommends two different ways: 

>   * exchanging cat()/print() with other generics
>     - [message()](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/message): for information messages and status updates 
>     - [warning()](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/warning): for warning, will print a "Warning: " before the output
>     - [stop()](https://www.rdocumentation.org/packages/base/versions/3.6.2/topics/stop): for error messages, will print an "Error: " before the output and halt the execution

Reviewer: CRAN still doesn't provide html manual pages of packages (+base) but it might be better to directly link to those in the r-project.org url: https://search.r-project.org/R/refmans/
Or from Martin Mäechler server: https://stat.ethz.ch/R-manual/R-devel/library/base/html/stop.html

```suggestion
  * exchanging `cat()`/`print()` with other generics
    - [`message()`](https://search.r-project.org/R/refmans/base/html/message.html): for information messages and status updates 
    - [`warning()`](https://search.r-project.org/R/refmans/base/html/warning.html): for warning, will print a "Warning: " before the output
    - [`stop()`](https://search.r-project.org/R/refmans/base/html/stop.html): for error messages, will print an "Error: " before the output and halt the execution
```

Reviewer 2: I got a "Forbidden: You don't have permission to access this resource." when clicking the link https://search.r-project.org/R/refmans/

> This allows to use functions like suppressMessages() to avoid unwanted output.  

Reviewer: ```suggestion
This allows to use functions like `suppressMessages()` to avoid unwanted output.  
```

  * using an additional function argument
    - create one argument in your function to turn off the console output

>     - CRAN suggests using `verbose`, other names are accepted

Reviewer: Perhaps mention the use of loggers too, as there are packages that implement that.
